src/main/java/com/github/swrirobotics/bags/BagService.java

```diff
@@ -763,6 +819,54 @@ private void addTopicsToBag(final BagFile bagFile,
         }
     }
 
+    @Transactional
+    public void addTagsToBag(final BagFile bagFile,
+                              final Bag bag) throws BagReaderException {
+        myLogger.trace("Adding tags to " + bagFile.getPath());
+        Set<Tag> bagTags = getTags(bagFile, bag);
+        List<Tag> dbTags = myTagRepository.findByBagId(bag.getId());
+        List<Tag> marked_for_removal = new ArrayList<Tag>();
+
+        // Delete old tags
+        for (Tag dbTag : dbTags) {
+            boolean found = false;
+            for (Tag bagTag : bagTags) {
+                if (Objects.equals(dbTag.getTag(), bagTag.getTag())) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                myLogger.debug("Removing old tag '" + dbTag.getTag() + "' which does not exist anymore.");
+                marked_for_removal.add(dbTag);
+            }
+        }
+        // TODO delete from database: Why is this not working? Fails with 'deleted instance passed to merge'
+        // myTagRepository.delete(marked_for_removal);
+        //for( Tag tag : marked_for_removal)
```

Comment:
I think there are two issues here.  The first is that I made a mistake in some of the persistence classes that I hadn't noticed until now because I had never tried deleting individual rows in any of the ancillary tables associated with bags. :smile:  In `Tag.java`, the `@ManyToOne` annotation on the bag member should look like this:
```java
    @ManyToOne(optional=false)
    private Bag bag;
```

In other words, it should *not* have the `cascade = CascadeType.REMOVE` property that is currently present.  That is causing Hibernate to attempt to delete the entire Bag record every time a Tag is deleted.  The BagPosition, MessageType, and Topic definitions all have similarly incorrect annotations.

Second, in addition to deleting the Tag instance from the repository, it also needs to be removed from the collection on the bag record.  The above lines should look like this:
```java
        myTagRepository.delete(marked_for_removal);
        bag.getTags().removeAll(marked_for_removal);
        bagRepository.save(bag);
        for( Tag tag : marked_for_removal) {
```

After making those changes, this code ran successfully for me.